Polish color indicator and itemgroup.

jasmussen opened the pull request:
## Description

This PR tweaks some of the metrics of the color swatch indicator to match those of the Global Styles design (#34574), which look like this in trunk:
<img width="296" alt="Screenshot 2021-12-01 at 11 31 49" src="https://user-images.githubusercontent.com/1204802/144223383-ff64502e-87b1-4439-a096-150a7b79893f.png">

The circle is a bit big, 24px, should be 20, and each row in the itemgroup is a bit big. This PR polishes a bit to this:

<img width="279" alt="Screenshot 2021-12-01 at 12 00 33" src="https://user-images.githubusercontent.com/1204802/144223450-dd511180-a482-4b28-8d1f-72890e83363c.png">

20x20px circles, about 40px height itemgroup rows. Note also how the gray border at the bottom goes edge to edge. There used to be a barely perceptible diagonal in the corners:

<img width="1197" alt="Screenshot 2021-12-01 at 11 49 31" src="https://user-images.githubusercontent.com/1204802/144223673-e8ad2d6e-b838-4259-b3cc-1e624ed1f466.png">

This was due to a transparent 1px border, adding the joint in the corner.

Note that there's some padding math in the itemgroup component that appears there to adhere to a component size system, so I didn't touch it. This actually makes the itemgroup rows still e a bit too tall (42.38px rather than 40px). I'm curious if there's a better way to solve this:

<img width="313" alt="Screenshot 2021-12-01 at 12 06 25" src="https://user-images.githubusercontent.com/1204802/144224003-09cec7cf-e1d6-467a-8324-ab1f17515229.png">

Also note that this PR completely moves the color indicator styles from global styles into the color circle component itself. As far as I can tell that component is used in the native code, and in global styles. But in both cases it should arguably be identical. I'd appreciate if someone familiar with the RN stuff could give this a quick sanity check.

## How has this been tested?

Test the color pickers in global styles. 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Polish color indicator and itemgroup.

jasmussen commented:
@sarayourfriend I think you've touched this code at some point. I know you're busy elsewhere so this is just an FYI, don't feel like you have to respond.

youknowriad commented:
I like that this tried to improve the root component :). This one might be used though in unexpected places (I think the title of the `PanelColorGradientSettings` component. (which might still be used in block sidebars). Does it still look ok there?

youknowriad commented on packages/components/src/item-group/styles.ts:
```diff
@@ -10,7 +10,7 @@ import { CONFIG, COLORS } from '../utils';
 export const unstyledButton = css`
 	appearance: none;
-	border: 1px solid transparent;
+	border: 0;
```
I wonder why this was using a transparent border

jasmussen replied:
I wonder the same. It could be related to vertical height/metrics. In any case removing it allowed the border to go edge to edge instead of stopping in a triangular joint.

ciampo replied:
This change currently breaks focus styles (see the `&:focus` rule a few lines below).

Leaving a transparent border allows us to only modify the `border-color` when the component receives focus, therefore avoiding an otherwise annoying "jump" that would be caused by adding a border

jasmussen commented:
> This one might be used though in unexpected places (I think the title of the PanelColorGradientSettings component. (which might still be used in block sidebars). Does it still look ok there?

Let me know if I'm looiking in the right place, but gradients look good to me:
![color](https://user-images.githubusercontent.com/1204802/144231333-77821cd6-f955-4858-82cd-f2e58e38c3c1.gif)

When I investigated which component to edit, I found these:

<img width="528" alt="Screenshot 2021-12-01 at 13 02 56" src="https://user-images.githubusercontent.com/1204802/144231365-4b6e1082-bfd2-4f94-9e33-c3e1b96f9ca2.png">

`ColorIndicator`, as far as my searching goes, is the only component that outputs the `component-color-indicator` classname. So yes, it appears to be used in some native stuff, but I'm not sure otherwise.

ciampo commented on packages/components/src/color-indicator/style.scss:
```diff
@@ -1,11 +1,8 @@
+	border: $border-width solid $gray-300;
 	display: inline-block;
-
-	& + & {
-		margin-left: 0.5rem;
-	}
```
This change could cause some layout glitches in other parts of Gutenberg

jasmussen replied:
I haven't found any places where this rule would apply. As is, I removed it because it appears to be dead CSS.

ciampo replied:
I'm not sure it's dead CSS — this rule would potentially affect those situations where 2 or more `ColorIndicator`s are rendered next to each other (at least in DOM order).

I scanned the codebase and found a few places that we should check to make sure we didn't introduce a layout regressions:

- [`PanelColorGradientSettings`](https://github.com/WordPress/gutenberg/blob/26556937b70bbf4aaef8c112321703b61a740284/packages/block-editor/src/components/colors-gradients/panel-color-gradient-settings.js#L75)
- Global Styles' [`Palette`](https://github.com/WordPress/gutenberg/blob/233e5b2fa55363a0343f865b697c7923d49eb017/packages/edit-site/src/components/global-styles/palette.js#L66-L69) (although this may be already tackled by the changes to `packages/edit-site/src/components/global-styles/style.scss` in this same PR)
- Global Styles' [`Preview`](https://github.com/WordPress/gutenberg/blob/251f00f45688fdb25e883e861b5ac788d58ec84b/packages/edit-site/src/components/global-styles/preview.js#L35-L36) (although this may be OK because of the `VStack` element wrapping around them)

ciampo reviewed (commented):
Hey @jasmussen , thank you for working on these changes!

For future changes to the `@wordpress/components` package, it would be better if you could open one PR per component (for example, in this PR we're introducing changes to both `ColorIndicator` and `ItemGroup`).

Regarding `ColorIndicator` in particular, some time ago we planned to explore creating a new `ColorSwatch` component, which would replace both `ColorIndicator` and `CircularOptionPicker` (more details in the `ColorSwatch` section of #35093), but unfortunately we currently lack the capacity to work on that.

> When I investigated which component to edit, I found these:

We should check each instance of `ColorIndicator` and see if we can remove any style overrides (similarly to what we've done in this PR in `packages/edit-site/src/components/global-styles/style.scss`) or if we need to add any fixes after the changes in the component.

Finally, would you mind adding 2 [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md) entries for `ColorIndicator` and `ItemGroup` ?

Thank you!

geriux commented:
All good from the mobile editor side 👍 It uses its native variant for the component and styles.

jasmussen commented:
> Leaving a transparent border makes allows us to apply a border-color change for focus styles, without having the component change its outer dimensions.

I'll revisit this piece and per your feedback to keep component PRs separate, probably restore that rule so this only touches the color indicator. Thank you!

jasmussen committed:
Revert itemgroup changes, add changelog entry.

jasmussen commented:
Thanks again for the feedback. I went ahead and reverted the ItemGroup changes (I'll revisit separately), and added a changelog entry for the colorindicator. Let me know if that looks good.

youknowriad commented:
<img width="282" alt="Screen Shot 2021-12-01 at 2 47 44 PM" src="https://user-images.githubusercontent.com/272444/144245782-f6e9263c-a956-42f0-976c-172ec1766808.png">

This is where this PR falls short, we can update that panel title though to the new styles of the indicator though.

jasmussen commented:
Oh that's excellent, thanks for bringing my attention to that. I think I can polish that up! That part needed a little love anyway, don't you think?
<img width="411" alt="Screenshot 2021-12-01 at 14 53 27" src="https://user-images.githubusercontent.com/1204802/144246605-a4f49db0-3ec5-4d78-8cb1-8885d61225ce.png">

I think with the right positioning or margin, that little circle will be better.

ciampo reviewed (commented):
> Thanks again for the feedback. I went ahead and reverted the ItemGroup changes (I'll revisit separately), and added a changelog entry for the colorindicator. Let me know if that looks good.

Speaking of which, I think that there are a few changes introduced in this PR that can definitely cause layout regressions (as Riad already pointed out).

Apart from [this conversation about the sibling selector](https://github.com/WordPress/gutenberg/pull/37028/files#r760129337), the fact that we're changing the `height` and we're removing the `margin-left` will definitely cause layout some layout regressions (as Riad already pointed out).

We should look for every instance of this component (I like to search for both the `ColorIndicator` _and_ the `component-color-indicator` strings) and make the necessary tweaks, mostly:
- add a `margin-left` when needed
- tweak vertical alignment and margin top/bottom (due to the change in height)

I know it can be quite a tedious task, apologies in advance! But I feel like these are definitely some good improvements — It's good that we're moving away from having a component that sets its own margin (which for me is an antipattern)

jasmussen committed:
Update collapsed panel indicator.

jasmussen commented:
Alright, I rebased and pushed tweaks based on your feedback. The global styles interface still looks like this:
<img width="277" alt="Screenshot 2021-12-02 at 12 07 55" src="https://user-images.githubusercontent.com/1204802/144410964-8ef24105-8e17-4075-8b0c-3880ab6ba354.png">

Now the dot in the collapsed panel looks like this:
<img width="410" alt="Screenshot 2021-12-02 at 12 07 03" src="https://user-images.githubusercontent.com/1204802/144410982-db44e7a6-4c84-4894-a339-86b4caef11c3.png">

I ended up restoring the left margin, as it did appear to be used for the color indicator inside the collapsed panel. But the sibling selector still appears unnecessary. Here's the pertinent bit:

```
 .component-color-indicator {
	...
	margin-left: 0.8rem;
	...
	& + & {
		margin-left: 0.5rem;
	}
}
```

Those margins should both be the same. So I just kept the simple margin and changed it to use the grid-units.

One thing to consider is to not let the age of this component hold us back from cleaning things up. In my digging I found that it appears to have been used for color swatches in #7924. But those now appear to use the `circular-option-picker` component instead. I also found reference to the component being used in the inline color UI: #16014. But as far as I can tell, that component isn't even used at all anymore, am I right in that?

<img width="521" alt="Screenshot 2021-12-02 at 12 00 32" src="https://user-images.githubusercontent.com/1204802/144411352-915242a5-9472-4633-9d54-74d9f5e31fd6.png">

The inline color picker looks like this now:
<img width="575" alt="Screenshot 2021-12-02 at 11 59 44" src="https://user-images.githubusercontent.com/1204802/144411470-d38c3534-203d-4124-8c5c-9a49ced41161.png">

That's the circular option picker component in use there as well.

What do you think?

youknowriad commented:
I think this all makes sense but for me if we can, we should remove the margin as it's not the responsibility of a component to add margin to position itself. It can be used in a lot of variety of places and a margin is specific to one use-case and should move to that place instead.

ciampo commented:
> I ended up restoring the left margin, as it did appear to be used for the color indicator inside the collapsed panel. But the sibling selector still appears unnecessary
> [...]
> Now the dot in the collapsed panel looks like this:

This sounds sensibile, thank you! I definitely trust your judgement when it comes to polishing the UI — if you're happy with the way it looks, then it's good for me too!

> But as far as I can tell, that component isn't even used at all anymore, am I right in that?

I'm not sure about `InlineColorUI` — since it is a default export, it means that the component itself could be renamed when imported. But still, it looks like `InlineColorUI` internally renders a `ColorPalette`, which internally renders a `CircularOptionPicker` (as you also confirmed).

> One thing to consider is to not let the age of this component hold us back from cleaning things up

Absolutely! I hope my comments are not being seen as too much "holding back" — I am just trying to make sure that we don't introduce breaking changes and layout regressions. I would also like very much to, for example, remove those margins (I believe that a component should _never_ set an outer margin on itself), but unfortunately these components have been around for a while and it's not easy to introduce breaking changes with the current state of Gutenberg and how packages are versioned / released.

_Edit: I agree with @youknowriad's comment — I think that the right solution would be to remove the margin rule from the `ColorIndicator`, and add it where it makes sense in the places where `ColorIndicator` is used._

As I mentioned before, we did have a [roadmap for the changes that we wanted to apply to Color-related components](https://github.com/WordPress/gutenberg/issues/35093), but unfortunately the components squad currently has very limited resources and no time to work on that. The fact that we decided to have the Global Styles sidebar in the upcoming WordPress release also shifted our focus on other lines of work, and therefore we didn't manage to make enough progress on this front.

I'd love it if we could get some collaboration of moving the work on color components forward — and working on a new `ColorSwatch` component would be one of the first tasks. It would replace (clean up!) both `ColorIndicator` and `CircularOptionPicker`, and then we could work our way up in refactoring/improving higher-level components (like palettes, gradient pickers...)

youknowriad commented:
I think some small styles "breakage" is fine if we go into the right direction. That margin there is especially obvious to me that it shouldn't be part of the component and if there's a breakage because of that (outside Gutenberg core since we'll be fixing all of that), I think it's small enough to be ok.

jasmussen commented:
Thanks all, I'll take another look at the margin and see if I can fix the case where you have both text and background colors defined, and how that shows in the collapsed panel. That seemed to be the primary place that margin was used.

jasmussen committed:
Use gap instead of margin.

jasmussen commented:
Okay, try the last one, it uses gap instead of margin:

<img width="1119" alt="Screenshot 2021-12-02 at 13 52 27" src="https://user-images.githubusercontent.com/1204802/144425758-fea59cc3-8825-47a9-9891-604afaf47205.png">

youknowriad commented on packages/block-editor/src/components/colors-gradients/style.scss:
```diff
@@ -7,6 +7,7 @@
 .block-editor-panel-color-gradient-settings {
 	.block-editor-panel-color-gradient-settings__panel-title {
 		display: flex;
+		gap: $grid-unit-15;
```
👍

youknowriad commented:
The circles do look too small to me on that panel thing but I guess it's just taste :)

ciampo reviewed (approved):
Thank you @jasmussen for polishing the `ColorIndicator`! 🚀

jasmussen commented:
Thanks both. I'll keep an eye on the php unit test update, rebase, and land this.

Regarding the small color dots, we could make them stack horizontally like the GS presets.

youknowriad commented:
Ignore the php unit failure for now, it's a bug introduced in trac/core, they should be working on it.

jasmussen commented:
Alright, I've got passing unit test, PHP tests are failing temporarily. So I'll ignore those for now. Thanks again folks!
